Better support for literals in jit script

zou3519 opened the pull request:
Addresses #8177

A design doc can be found here: [gist](https://gist.github.com/zou3519/4b7f13f03cc9f3612bd9363e6405fa0a) version or [quip](https://fb.quip.com/azL1AqUckBdo) version

General approach:
- Add NumberType, FloatType, IntType to represent Python numbers, floats and ints.
- Emit these types for python literals
- Change aten_schema such that Scalars are NumberType, int64_t and bool are IntType.
- Emit aten::type_as, prim::NumToTensor, and prim::TensorToNum nodes for tensor-number math. (see examples below)
- Erase NumberType,  prim::NumToTensor, and prim::TensorToNum for ONNX export. Let ONNX export the aten::type_as nodes.

### Tensor/number math
```
import torch
@torch.jit.script
def fn(x):
    return x + 1
```
```
graph(%x : Dynamic) {
  %1 : int = prim::Constant[value={1}]()
  %2 : Dynamic = prim::NumToTensor(%1)
  %3 : Dynamic = aten::type_as(%2, %x)
  %4 : Dynamic = aten::add[alpha={1}](%x, %4)
  return (%5);
}
```

### Number/Number Math
```
import torch
@torch.jit.script
def fn(zero):
    c = 1 + 1
    return zero + c
```
```
graph(%zero : Dynamic) {
  %1 : int = prim::Constant[value={1}]()
  %2 : int = prim::Constant[value={1}]()
  %3 : Dynamic = prim::num_to_tensor(%1)
  %4 : Dynamic = prim::num_to_tensor(%2)
  %5 : Dynamic = aten::add[alpha={1}](%3, %4)
  %c : int = prim::TensorToNum(%6)  # this is the result of the addition
  ...
  return (%13);
}
```

cc @zdevito @jamesr66a 

There are a list of follow up tasks that I will file issues for later. If any of these seem important right now I can fold them into this PR.
- Let's say I called prim::NumToTensor on an int. Should the output be of type Long()? (It is currently DynamicType)
- Allow scalar tensors for Scalar args
- Add bool type to the schema to differentiate between int and bool
  - fix var/std
- Test logical operators and make them follow python semantics
- Support passing in non-tensor arguments to script functions/support returning non-tensor values from script functions
- Change the schema for IntList to List[int] or something similar
- Better support for List[int] or Tuple[int] args
- Add peephole to delete unnecessary type_as. This peephole should be probably be run on specialized graphs (in an unspecialized graph, it is hard to tell if a type_as is unnecessary).
- we should clean up the math handling code. Instead of using if statements on types, we can just register additional builtin ops like __add__ and __le__. Including __rdiv__ style ops to handle the reverse cases. This will make it easier to see how math is getting emitted and ensures the math functions always go through the same type-conversions as built in functions.

zou3519 committed:
Introduce Python Number types

Added: IntType, FloatType, NumberType with
IntType <: NumberType
FloatType <: NumberType

Changed aten_schema so arguments have corresponding types

zou3519 committed:
Emit a NumberType for python literals.

Also emit a NumberType for Scalar default values.

zou3519 committed:
Add prim::NumToTensor and prim::TensorToNum

zou3519 committed:
Add DynamicType -> NumberType implicit cast for bc

zou3519 committed:
Better ensureTensor error message

zou3519 committed:
Add ensureTensorOrNumber. Allow passing Number to some functions

Like the range() construct and slices

zou3519 committed:
Patch IntList to work.

IntList is still a DynamicType in the frontend: a tensor gets built from
a List[int].

Also, IntList[1] is a "union between int and IntList" the way it is
implemented. If the frontend sees an int being passed for an IntList[1]
arg, it converts it to a tensor as well.

zou3519 committed:
Enforce some order on schemas to avoid overload ambiguity

add(Tensor, Tensor) should appear earlier than add(Tensor, Scalar). This
matches the order in which python_arg_parser parses its arguments.

zou3519 committed:
Disable std_dim and var_dim tests.

With the new schema information, std(input, keepdim) and std(input, dim)
are ambiguous. This will need to be fixed at a later date.

zou3519 committed:
Add NumberType erasure pass.

This is used for ONNX export and to ensure that NumberType information
doesn't reach the interpreter

zou3519 committed:
Add support for mixed tensor/number math ops.

zou3519 committed:
Tests for new functionality.

Includes:
- Tensor/number math
- number/number math
- EraseNumberTypes pass test

zou3519 committed:
Patch tests

Update expect tests for:
- decompose_addmm
- loop unrolling tests

Because python numbers are now NumberType, they cannot be returned by
functions anymore. Work around this by using "torch.full", or by adding
a tensor([0]) (taken from FIXME_zerol()). Both approaches are used
because torch.full is more readable, but it is broken in some cases.

zou3519 committed:
Add erase_number_types to torch/CMakeLists.txt

zdevito commented on torch/csrc/jit/graph_executor.cpp:
```diff
@@ -294,6 +295,7 @@ struct GraphExecutorImpl {
     auto local_graph = this->graph;
     if(all_dynamic(local_graph->inputs()) && all_dynamic(local_graph->outputs())) {
       local_graph = this->graph->copy();
+      EraseNumberTypes(local_graph);
```
Is this here because PropagateInputShapes does not understand number types? Can we add a comment if so?

zou3519 replied:
Yes

zdevito commented on torch/csrc/jit/interned_strings.h:
```diff
@@ -29,6 +29,11 @@ enum class SymbolNamespace : unsigned char {
 //
 // TODO: Consider moving the synthetic onnx operators to their own
 // namespace.
+//
```
This seems dead.

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -108,6 +109,7 @@ struct Method {
     for (auto inp : member_inputs) {
       inputs.push_back(*inp);
     }
+    EraseNumberTypes(retval);
```
Ok, I don't like how in how many places this is appearing. This also looks like it is because PropagateInputShapes does not work with number types. I think the right fix is to add handling to PropagateInputShapes for the operators that you added. This can be done in a follow up PR, since we are blocked on other things waiting for number type.

zou3519 replied:
Will add to the list of followup prs

zou3519 replied:
Oh, this one here in particular is to patch ONNX export, I'll try to see if there's a better place to put it.

zdevito commented on torch/csrc/jit/test_jit.cpp:
```diff
@@ -870,6 +873,8 @@ void testControlFlow() {
   auto run = [&](const std::string & name, std::vector<at::Tensor> stack) {
     auto graph = cu.get_method(name).graph();
+    // TODO: why isn't this (a required pass) called?
+    EraseNumberTypes(graph);
```
We should decide _when_ EraseNumberTypes happens. If it is just so the interpreter doesn't have to handle it then the `Code` constructor should be the _only_ one to call it, and the calls in the graph executor should be removed.

zdevito commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -274,6 +274,32 @@ def emit_decl(decl):
         for variant in {all_real_arguments_are_inputs, only_tensors_are_inputs}:
             emit_decl_variant(decl, variant, has_tensorlist)
 
+    # This function declares an order on declarations. This is necessary because
```
This should become unnecessary when we remove the implicit conversion from Tensor -> Number, right? We should make a note of it here, because it is quite complicated.

zou3519 replied:
Most likely yes. There's an edge case that I'm worried about (pytorch allows one to pass in a scalar tensor as both a Scalar and Tensor argument, and if we allow that too in jit script the ordering will matter), but I think Sam is working on getting rid of overloads like this. I'll make a note here.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1227,6 +1488,27 @@ struct to_ir {
       case TK_GE:
       case '*':
       case '/':
+      case '+':
```
Why did these move into emitSugaredExpr? They are still always returning values, which was the purpose of the emitSimpleExpr node.

zou3519 replied:
Good point, I'll move them back into the emitSimpleExpr path

zdevito reviewed (approved):
This looks good. The follow up issues that you noted will need to be done, but we can parallelize them across multiple people, so I want to get this merged. The only things that should be fixed are:
* Ensure we are not breaking ONNX tests with this change, in particular, I see an ATen op in one of the export strings.
* Move the math handling code back to emitSimpleExpr unless there is a reason it cannot go there. 

I would also add to the task list: we should clean up the math handling code. Instead of using if statements on types, we can just register additional builtin ops like `__add__` and `__le__`. Including `__rdiv__` style ops to handle the reverse cases. This will make it easier to see how math is getting emitted and ensures the math functions always go through the same type-conversions as built in functions.

zou3519 commented on test/test_jit.py:
```diff
@@ -3639,6 +3808,14 @@ def forward(self, x, y):
     'test_clamp_max_scalar',
```
I haven't checked, but I'll make a pass through before this is merged

apaszke replied:
We should probably try to run those tests as expected failures, just to make sure they won't sit there forever for no good reason

zou3519 committed:
Move math back to emitSimpleExpr from emitSugaredExpr

zou3519 committed:
Remove some dead lines

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -434,15 +490,37 @@ at::optional<std::vector<Value*>> tryMatchSchema(
-      // implicit conversion from List[Tensor] -> Tensor for when the argument
+      // An IntList[1] is a union of int and IntList.
+      // Reinterpret an int as an "IntList" (a tensor)
+      if (isIntUsedAsIntList(v.value, arg)) {
```
Oh god, why? This surely deserves an explanation

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -434,15 +490,37 @@ at::optional<std::vector<Value*>> tryMatchSchema(
 
+      // implicit conversion from Tensor to Python Number
+      // FIXME: remove this when we support passing numbers into script fns
+      if (v.value->type()->isSubtypeOf(*DynamicType::get()) &&
```
Use `isTensorSubtype`? It's much more readable than this expression

zou3519 committed:
Renable some excluded script/trace tests that are fixed.

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1188,6 +1276,187 @@ struct to_ir {
+    for (Value* inp : inputs) {
+      result.push_back(NamedValue(loc, "", inp));
+    }
+    return result;
```
nit: use `fmap` (does e.g. `reserve` which you forgot here)

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1227,6 +1488,27 @@ struct to_ir {
+      const TreeRef& tree) {
+    switch (tree->kind()) {
       case '@':
       case TK_POW:
```
Shouldn't we treat POW specially as well?

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1237,14 +1519,6 @@ struct to_ir {
         auto kind = getNodeKind(tree->kind(), inputs.size());
```
Should the UNARY_MINUS stay here? Will this case ever get triggered?

zou3519 replied:
Will change

zou3519 replied:
Resolved; I moved the other UNARY_MINUS case to this location so that this one is the only one triggered.

apaszke commented on test/test_jit.py:
```diff
@@ -1226,7 +1232,9 @@ def func4(a, b):
-        self.checkScript(func4, (a, b), optimize=True)
+
+        # TODO: renable when we support passing literals to script fns
+        # self.checkScript(func4, (a, b), optimize=True)
```
Can you make it an expected failure?

apaszke commented on test/test_jit.py:
```diff
@@ -52,6 +52,12 @@
 
 
+# TODO: Replace all uses of this function with the literal "0" when the jit
+# is able to support returning numbers (as opposed to only Tensors)
```
What's blocking us from doing it? The fact that they would be scalar tensors and not Python numbers?

zou3519 replied:
Two things:
1) Yes, the fact that they would be scalar tensors and not Python numbers
2) There is a check somewhere that the result of a function is a tensortype. A python number has type NumberType so it gets caught by this check. I haven't gotten around to seeing what happens if I disable that check yet.

apaszke commented on test/test_jit.py:
```diff
@@ -1421,18 +1429,20 @@ def func(lim):
+                a = zero + 3 + d
             else:
-                b = 3 + d
+                b = zero + 3 + d
```
Can we please avoid reusing existing tests, with well-defined scope for checking this extra behavior? If I were to refactor the test in the future, it's possible I would remove this stuff, because it doesn't contribute to `test_if` in any way

zou3519 replied:
The reason why I added this in here is that `a` and `b` are both treated as DynamicType. `3 + d` is a IntType. We disallow assigning DynamicType a IntType value, so I had to convert the IntType value to a DynamicType value somewhere.

apaszke replied:
Ugh, we should probably relax those constraints. If you can't change tensors to numbers, natural things like
```python
loss = 0.
for o in outputs:
    loss += o.sum()
```
will break.

apaszke replied:
Actually can we just fix the broken tests by removing the scalar literals (use tensor math only)? The fact that so many of them needed to be change is a sign that our current strategy is way too strict

zou3519 replied:
Do you mean changing all of the python numbers to scalar tensors, or in general just getting rid of python numbers in the tests and sticking in tensors somehow?

apaszke commented on test/test_jit.py:
```diff
@@ -1509,17 +1555,19 @@ def func(a, b):
     def test_script_for_in_range_dynamic(self):
         def fn():
-            c = 0
+            c = FIXME_zerol()
```
Is the `0` literal so broken right now that we can't even use it? Why?

zou3519 replied:
Yes, the 0 literal is broken for a number of reasons. In this particular case the problem is that `for i in range(100)` assigns `i` a TensorType. In the above test (test_script_for_in_range), we're using `c` as the accumulator and adding `i` to `c`.

If we assign `c = 0` originally, `c` has type IntType. However, `i` has some tensor type, and we disallow reassigning variables to things that have unrelated types. This can be fixed when the `range(...)` expression gives an IntType.

I can add more comments to the code if you think that is fine

apaszke commented on test/test_jit.py:
```diff
@@ -2538,9 +2692,14 @@ class M(torch.jit.ScriptModule):
+            # TODO: Use this when we support passing literals to script fns
+            # @torch.jit.script_method
+            # def call_foo(self, input):
+            #     return self.foo(input=input, bar=1)
```
Expected failure pls

apaszke commented on test/test_jit.py:
```diff
@@ -3639,6 +3808,14 @@ def forward(self, x, y):
+    'test_var_dim',
+    'test_var_dim_1d',
+    'test_var_dim_1d_neg0',
+    'test_var_dim_neg0',
```
Why did those start to fail now?

zou3519 replied:
I think what's going is the following:

there are two schemas for `var`: (* means that after this point everything is a kwarg).
(1) var(Tensor, int, *, bool, bool, Tensor)
(2) var(Tensor, *, bool)

(1) used to come before (2) in aten_schema.cpp.

I made the following changes to schemas:
- Shorter schemas come before longer schemas
- bool, int are treated as IntType rather than DynamicType like before

So the schemas look like the following in aten_schema:
(2) var(DynamicType, IntType)
(1) var(DynamicType, IntType, IntType, DynamicType)
Now, when you call `torch.var(tensor, dim=1)`, the compiler mistakingly matches it with (2) instead of (1), which is a problem.

I'll put a comment in the code with this (and think about if I can fix this quickly or not)

zou3519 committed:
Move some tests to expected failure

zou3519 committed:
Address some comments (more addressing to come)

jamesr66a commented on test/expect/TestScript.test_onnx_export_script_module_if.expect:
```diff
@@ -6,20 +6,21 @@ ModelProto {
         Node {type: "Constant", inputs: [], outputs: [2], attributes: [{ name: 'value', type: tensor, value:TensorProto shape: []}]},
-        Node {type: "Greater", inputs: [1,2], outputs: [3], attributes: []},
-        Node {type: "If", inputs: [3], outputs: [4], attributes: [{ name: 'then_branch', type: graph, value:
+        Node {type: "ATen", inputs: [2,1], outputs: [3], attributes: [{ name: 'operator', type: string, value: 'type_as'}]},
```
Yes, I believe so. The form `expr > 0` seems pretty common and would be expected to work for anyone used to implicit conversions. Can we solve this restricted case by doing a peephole optimization that matches (Constant -> type_as) and cast the Constant node to the expected type? If `expr` is dynamic, we're probably screwed though

zou3519 replied:
Thanks for the confirmation, @jamesr66a.  For now I think I could eliminate the aten::type_as nodes that I added right before exporting to onnx so that the behavior before this PR is unchanged.

ezyang commented:
@pytorchbot retest this please

zou3519 committed:
Erase relevant aten::type_as nodes in EraseNumberTypes

I also changed it so that EraseNumberTypes is only called for ONNX
export. It is no longer used to prevent
prim::NumToTensor/prim::TensorToNum from reaching shape_analysis or
interpreter.cpp.

shape_analysis infers the type of the output of these nodes to be the
same as their input.

intepreter.cpp treats both of these nodes as no-ops.

zou3519 commented:
I've addressed the two points that @zdevito brought up:

1) The code doesn't break ONNX export anymore: the EraseNumberTypes pass erases aten::type_as nodes that are added for math operations. I changed it so that EraseNumberTypes is only called in the ONNX export path; shape analysis and interpreter.cpp can handle the casting operations now.
2) Moved the math handling code back to emitSimpleExpr

In addition, I took a look at the tests we skipped. [Here is a list of the ones that are fixed by this PR]( https://gist.github.com/zou3519/fdcaaa86eaad69f8b8a2e6a8d4ab9539) (I've renabled those tests.

zou3519 committed:
Add reminder to fix std/var

zou3519 committed:
Call EraseNumberTypes only when exporting a script module

zou3519 commented on torch/onnx/utils.py:
```diff
@@ -167,6 +167,8 @@ def _model_to_graph(model, args, f, verbose=False, training=False,
             graph = method.propagate_and_assign_input_and_output_shapes(
                 args, example_outputs, False, propagate)
+            # Erase number types to bring the graph to a pre-NumberType state
+            torch._C._jit_pass_erase_number_types(graph)
```
@jamesr66a does this look like the right place to put this?

zou3519 committed:
Update expects after rebase

zou3519 commented:
caffe2 test failure looks unrelated

jamesr66a reviewed (approved):
k

zou3519 commented:
Going to merge this PR now to unblock @zdevito and the robustness work being done by @jamesr66a and @wanchaol.

@apaszke I can address your comments about the tests in a followup PR (let me know if you would like me to change the tests to use tensors instead of scalars, or if we should consider allowing assigning tensors to ints, like in the common case you mentioned)

apaszke commented:
Let's just merge it now, and talk about relaxing the type constraints in the next meeting
